miden-crypto/src/merkle/smt/historical/mod.rs

```diff
@@ -0,0 +1,500 @@
+                        // All old values were empty, so the leaf was empty before
+                        SmtLeaf::Empty(leaf_index)
+                    } else if old_key_values.len() == 1 {
+                        // Single old key-value pair
+                        let (key, value) = pre_apply;
+                        if value != EMPTY_WORD {
+                            // The key had a non-empty value before, so it was a single entry
+                            SmtLeaf::Single((key, value))
+                        } else {
+                            // The key had an empty value before, so the leaf was empty
+                            SmtLeaf::Empty(leaf_index)
+                        }
+                    } else {
+                        // Multiple old values - filter out any with EMPTY_WORD values
+                        let entries = Vec::<(Word, Word)>::from_iter(
+                            old_key_values.into_iter().filter(|(_, v)| *v != EMPTY_WORD),
+                        );
+                        SmtLeaf::new(entries, leaf_index)
+                            .expect("consistency is given by construction")
+                    }
+                },
+                SmtLeaf::Single((current_key, current_value)) => {
+                    // Current leaf has a single entry
+                    // Check if this entry was modified
+                    let mut leaf_entries = vec![(current_key, current_value)];
+
+                    // Replace with old values
+                    for (old_key, old_value) in &old_key_values {
+                        if *old_key == current_key {
+                            // Replace the current value with the old value
+                            leaf_entries[0].1 = *old_value;
+                        } else if *old_value != EMPTY_WORD {
+                            // This key was removed in the mutation, add it back
+                            leaf_entries.push((*old_key, *old_value));
+                        }
+                    }
+
+                    // Filter out empty values and construct the appropriate leaf type
+                    let entries = Vec::<(Word, Word)>::from_iter(
+                        leaf_entries.into_iter().filter(|(_, v)| *v != EMPTY_WORD),
+                    );
+
+                    SmtLeaf::new(entries, leaf_index).expect("consistency is always given")
+                },
+                SmtLeaf::Multiple(current_entries) => {
+                    // Current leaf has multiple entries
+                    let mut leaf_entries = current_entries.clone();
+
+                    // Apply the old values
+                    for (old_key, old_value) in &old_key_values {
+                        // Find if this key exists in current entries
+                        if let Some(pos) = leaf_entries.iter().position(|(k, _)| k == old_key) {
+                            if *old_value != EMPTY_WORD {
+                                // Update the value
```

Comment:
Consider fine-grained locking for better concurrency. Current Arc<RwLock<InnerState>> creates contention between historical views on different tree branches. A Pareto-efficient approach using parking_lot::RwLock<Word> per node could provide 80% concurrency improvement with minimal code changes.